### ISSUE TITLE
Encrypted secrets in actions

### ISSUE THREAD

**Issue by lucie-docs** [githubtraining/lab-use-secrets#18]
Hi @mattdavis0351 and @hectorsector,

Just wanted to make you aware of a note I added in the **Manage GitHub Actions in the enterprise** MS Learn module: 

![Screenshot 2021-06-09 at 10 02 52](https://user-images.githubusercontent.com/40564413/121316784-f78c7c80-c909-11eb-8c8d-d6e430897fc9.png)

I added it following feedback from @dmckinstry who explained to me that using encrypted secrets in actions is considered bad practice and that it should probably be made clear in the content. His feedback also reminded me of [this comment](https://github.com/github/programs-forge/issues/159#issuecomment-849007878 ) from @djredman99 when you two were gathering feedback on the exercise.

I'm thinking it might be weird to have an exercise in the MS Learn module which asks the learner to pass a secret in an action when we've just told them not to do that or only as a last resort 😅 

Do you have any thoughts on this? I know we already added a note in the content of this exercise to explain how the PAT is used + documentation about the action but I can't help but think that we're going to end up advertising a bad practice.  Maybe we could reframe the exercise in a future version?

CC @a-a-ron for awareness

**Comment by lucie-docs** [comment]
Some more context around this I just remembered: another Xpirit SME reminded us in an interview with other content writers of the security risks linked to actions in general such as bitcoin-miners, etc. We know GitHub Actions is very powerful and that anything GitHub puts out there can be misused (also see [this case](https://github.com/github/learning-lab/issues/2686) I once had when doing comms duty). 

So I'm thinking we might not wanna train people on a bad practice that can lead to security issues?

**Comment by brianamarie** [comment]
Just to be sure I'm understanding the situation, here are my impressions (please correct as needed):
- The current plan is to use the new Secrets exercise
- We have received feedback from SMEs, both on the exercise alone and the content, that this usage of putting a secret in an Action is risky and not recommended (where as secrets in the WORKFLOW are recommended)

If that's all correct, I see the following as our options (again, please correct as needed):
- We keep the Secrets exercise as is (with a note as @lucie-docs mentioned) despite SME guidance from multiple people at multiple stages that this is NOT what we want to be encouraging
- We do not use this secrets exercise until it has been amended to align with recommended practices
  - Option 1: We point instead to existing exercises (Learning Lab - @lucie-docs, you mentioned a we have an existing Learning Lab that does cover secrets?) 
  - Option 2: Similarly to our upcoming Admin content, we forego a hands-on exercise at this step for another, less-interactive validation

My two cents: I am of the opinion that shipping exercises that require the users to execute practices that we would not recommend to our customers has the potential for more harm than good. 

@a-a-ron I would like you to make a decision on how to move forward. Even with my two cents above, I believe you are closest to the content, the quality, and the audience, and I will defer to you on this. Is that OK with you @a-a-ron?

**Comment by lucie-docs** [comment]
> Option 1: We point instead to existing exercises (Learning Lab - @lucie-docs, you mentioned a we have an existing Learning Lab that does cover secrets?)

Yes, [this one](https://lab.github.com/githubtraining/github-actions:-continuous-delivery-with-azure) uses secrets in workflows to deploy to Azure.

**Comment by mattdavis0351** [comment]
> > Option 1: We point instead to existing exercises (Learning Lab - @lucie-docs, you mentioned a we have an existing Learning Lab that does cover secrets?)
> 
> Yes, [this one](https://lab.github.com/githubtraining/github-actions:-continuous-delivery-with-azure) uses secrets in workflows to deploy to Azure.

There is no other way to deploy to Azure. 

I'm not sure why our Actions docs would show people how to use encrypted secrets in actions if it was a bad practice. 

There are many verified actions on the marketplace, like the Azure actions, that require API keys as encrypted secrets or the don't work. I wouldn't begin to say that using my Azure API token with an Azure action is a bad practice. 

Where I think there is a misunderstanding is that if you are passing your secrets to "Joe's random action" then that's a bad practice. You should review every action that isn't verified before you pass it secrets just to be sure nothing bad is taking place. 

Lastly, if I'm an org and I have a local action in a private repo that means that someone who has access to that repo wrote the action. This means it's internal and not meant to be shared and it was built by people we trust. If we pass a secret to that action I wouldn't call that a bad practice, the entire thing is private and built by trusted devs. 

I don't think passing secrets is a black and white situation on whether it's good or bad.

**Comment by lucie-docs** [comment]
But in the Azure learning lab, the secret is passed at workflow level, not action level (I don't remember being asked to pass the secret in an action as a learner but maybe I'm wrong). I don't think I misunderstood what the SMEs told me, Dave said word for word "this is a bad practice" and could not think of any cases where you wouldn't have a way to put the secret in the workflow instead of the action. These verified actions on the marketplace might be there and valid but that does not mean this is what we want to teach to a learner first.

Also in the MS Learn module, this note I added in the content is literally the last thing they read before going into the exercise. We tell them "it's generally not recommended to use secrets in actions" and 3 seconds later we train them on the very specific scenario of passing a secret in an action. I'm just saying I don't think this is the best experience for the learner but I'll leave it up to @a-a-ron to weigh on this.

**Comment by lucie-docs** [comment]
> I'm not sure why our Actions docs would show people how to use encrypted secrets in actions if it was a bad practice. 

Just FYI, documentation teams generally document to the best of their knowledge and have to document everything that is possible with a product **at release**. Teaching good and bad practices is not central to documentation even if as a documentarian you should add that context if you're aware of it. But since good and bad practices are generally defined over time (i.e. when your product has been there long enough that people can identify these good and bad practices), it's unlikely that a docs team would be aware of them. In my opinion, this is where learning content comes into play, this is where people get more context about why and how you should use features. And this is why I think we should be featuring examples pointing learners in the direction we want them to take rather than exception scenarios or scenarios that require a deep knowledge of security and actions.

**Comment by dmckinstry** [comment]
I want to make sure we're not saying different things...  If I write and action and consume a secret, that is fine... as long as I'm consuming it as a parameter.  To the action, it is just a parameter like any other (secret or not).  So in this case, the secret is used in the workflow and (the way I was explaining) not in the action.  Does that make things clearer or more complex?

**Comment by lucie-docs** [comment]
That's the way I understood it @dmckinstry, maybe you can tell me if the note I wrote in the content reflects that correctly? 

Or maybe you can elaborate on what you referred to as bad practice to see if we're all talking about the same things?

**Comment by mattdavis0351** [comment]
> I want to make sure we're not saying different things... If I write and action and consume a secret, that is fine... as long as I'm consuming it as a parameter. To the action, it is just a parameter like any other (secret or not). So in this case, the secret is used in the workflow and (the way I was explaining) not in the action. Does that make things clearer or more complex?

That makes sense, that is exactly how we are doing it for the secrets exercise.  They pass the secret as a parameter to the local action.

>any cases where you wouldn't have a way to put the secret in the workflow instead of the action.

This is correct, none of the actions we are writing for these infer a secret exists without having it passed as a parameter outside of a standard `GITHUB_TOKEN`.  Anything other than that, which is a default secret, must be passed.


>(I don't remember being asked to pass the secret in an action as a learner but maybe I'm wrong)

You're sorta correct.  I wrote most of the workflow for you already, so therefore things like this existed in the workflow without you having to type it all out.  Here we are using the secret as an environment variable which get's used in multiple run steps later.
![image](https://user-images.githubusercontent.com/38021615/121399316-27884f80-c90b-11eb-8e7c-29fda8532cba.png)


And here we use an Azure specific action which takes a secret as a parameter
![image](https://user-images.githubusercontent.com/38021615/121399606-71713580-c90b-11eb-9795-98257a3f209c.png)

So we do this very frequently throughout that Learning Lab.  

What @dmckinstry said makes a lot of sense from how I'm inferring it, which is use secrets all you want as long as it's explicitly passed into the action by the user, rather than the source code of the action consuming the secret behind the leaners back. Which our team wouldn't author things this way anyway, but it's important to have on our radar when we invite outside authors in to help.  

We can easily check the `action.yml` file, along with whatever source code they provide to make sure that there is no hidden secret usage that isn't passed via parameter.  This should be a part of our approval process for accepting exercises from outside authors.

**Comment by mattdavis0351** [comment]
> We do not use this secrets exercise until it has been amended to align with recommended practices

@brianamarie This exercise already meets this criteria.  We are asking the learner to supply a secret to an action as a parameter to that action.  This is not a "bad" practice, rather this is the way we should be doing it.  They supply that secret through the workflow.

I think there was a misunderstanding about how the secret was being used/supplied that set this issue into motion.

**Comment by a-a-ron** [comment]
Thank you for this conversation, @mattdavis0351, @brianamarie, @lucie-docs, and @dmckinstry. 

From a content perspective, I don't want to include exercises in the MS Learn content that follow bad practices or at least give the impression of such even if we provide a note explaining why (unless that is our objective). This can be confusing. I appreciate everyone's comments and clarification to help resolve any misunderstandings.

From what I've gathered from recent conversations and going through the exercise again, the exercise doesn't seem to fall into "bad practices" because as mentioned above, it supplies a secret to the action as a parameter to that action. At this point, It seems that we can and should improve the leading content within MS Learn to add clarity into good/bad practices when working with secrets and then provide the appropriate context to introduce this exercise within the MS Learn unit. 

@lucie-docs, I'm reviewing this content today and i'll go through this particular part with these conversations in mind. I'd like a final review from a SME that is familiar with this exercise to approve the leading content before we include the exercise unit in the module. If our SMEs are concerned with the initial use of this exercises then we need to improve the content that supports this exercise.

**Comment by lucie-docs** [comment]
Aaah ok I see where the misunderstanding is. I did not misunderstand @dmckinstry and I think the content I wrote reflects what was explained in this thread. However, I did misunderstand what the exercise is asking the learner to do because I was _influenced by the content leading up to it_. Full disclosure, I only reviewed the instructions and did not do the exercise myself, maybe should have started with that LOL. 

However, some learners might be like me, only skim through instructions and misunderstand stuff, you cannot predict a learner's behavior (I mean learners like me, remember that time I disabled all workflows in one of the first hands-on lab? 😄). I think it makes sense to put content and exercise in front of a SME @a-a-ron just for peace of mind. And we can add/adjust the content according to the feedback!
